### PR TITLE
fix: storage mount

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

### Overview

The storage mount, even if `/proc/mount` returns the valid path, isn't quite mounted with proper permissions and the path is wiped out on mount. This change waits for the storage-ready hook to fire first, changing the permissions for the Jenkins home directory, making sure that Jenkins installation happens *after* the storage has been properly mounted.

### Rationale

When storage gets attached during jenkins installation (the /proc/mount will have /var/lib/jenkins) directory - Jenkins errors with file permissions error. This change will make sure that the `storage-mounted` hook will be handled before jenkins_pebble_ready event.

### Juju Events Changes

None.

### Module Changes

jenkins_storage_mounted is migrated from state to jenkins module.

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/jenkins.py [hunk 1]
```diff
@@ -158,6 +158,45 @@ def wait_ready(timeout: int = 300, check_interval: int = 10) -> None:
...
+        """Initialize a new instance of the StorageMountError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+def is_storage_ready(container: ops.Container) -> bool:
+    """Return whether the Jenkins home storage is mounted.
```
Comment (arturo-seijas) [inline src/jenkins.py:178]: Mounted and owned by jenkins?

File: src/jenkins.py [hunk 2]
```diff
@@ -158,6 +158,45 @@ def wait_ready(timeout: int = 300, check_interval: int = 10) -> None:
...
+
+    Returns:
+        True if storage is mounted, False otherwise.
+    """
+    mount_info: str = container.pull("/proc/mounts").read()
+    if str(JENKINS_HOME_PATH) not in mount_info:
+        return False
+    proc: ops.pebble.ExecProcess = container.exec(["stat", "-c", "%U", str(JENKINS_HOME_PATH)])
+    try:
+        (stdout, _) = proc.wait_output()
```
Comment (arturo-seijas) [inline src/jenkins.py:194]: I don't think you need parenthesis here

File: src/jenkins.py [hunk 3]
```diff
@@ -158,6 +158,45 @@ def wait_ready(timeout: int = 300, check_interval: int = 10) -> None:
...
+    """Return whether the Jenkins home storage is mounted.
+
+    Args:
+        container: The Jenkins workload container.
+
+    Raises:
+        StorageMountError: if there was an error getting storage information.
+
+    Returns:
+        True if storage is mounted, False otherwise.
```
Comment (arturo-seijas) [inline src/jenkins.py:187]: Same here

File: src/jenkins.py (hunk 1 above)
Comment (yanksyoon) [reply src/jenkins.py]: Updated, thanks!

File: src/jenkins.py (hunk 2 above)
Comment (yanksyoon) [reply src/jenkins.py]: Updated, thanks!

File: src/jenkins.py (hunk 3 above)
Comment (yanksyoon) [reply src/jenkins.py]: Updated, thanks!

**Review by arturo-seijas — APPROVED** [review]

**Review by javierdelapuente — APPROVED** [review]

**Pull request merged**